portal/frontend/src/main/kotlin/com/sourceplusplus/portal/page/TracesPage.kt

```diff
@@ -52,4 +115,428 @@ class TracesPage(
+                operationName = traceResult.artifactSimpleName.toString()
+            }
+
+            val rowHtml: HTMLTableRowElement = document.create.tr {
+                id = "trace-${htmlTraceId}"
+                td {
+                    onClickFunction = {
+                        clickedDisplayTraceStack(
+                            traceResult.appUuid,
+                            traceResult.artifactQualifiedName,
+                            globalTraceId
+                        )
+                    }
+                    style = "border-top: 0 !important; padding-left: 20px"
+                    i {
+                        style = "font-size:1.5em;margin-right:5px"
+                        classes = setOf("far", "fa-plus-square")
+                    }
+                    span {
+                        style = "vertical-align:top"
+                        +operationName.replace("<", "&lt;").replace(">", "&gt;")
+                    }
+                }
+                val occurred = moment(trace.start)
+                val now = moment()
+                val timeOccurredDuration = moment.duration(now.diff(occurred))
+                td {
+                    classes = setOf("trace_time", "collapsing")
+                    id = "trace_time_$htmlTraceId"
+                    attributes["data-value"] = trace.start.toString()
+                    style = "text-align: center"
+                    +getPrettyDuration(timeOccurredDuration, 1)
+                }
+                td {
+                    classes += "collapsing"
+                    +trace.prettyDuration!!
+                }
+                td {
+                    classes += "collapsing"
+                    style = "padding: 0; text-align: center; font-size: 20px"
+                    if (trace.error!!) {
+                        i {
+                            classes = setOf("exclamation", "triangle", "red", "icon")
+                        }
+                    } else {
+                        style.plus("color: #808083")
+                        i {
+                            classes = setOf("check", "icon")
+                        }
+                    }
+                }
+            } as HTMLTableRowElement
+            console.log(">>>>>>>>>> rowHtml = $rowHtml")
+
+            val traceTable: dynamic = document.getElementById("trace_table")
+            val tableRow = traceTable.rows[i]
+            if (tableRow != null) {
+                if (tableRow.id != "trace-$htmlTraceId") {
+                    traceTable.rows[i].outerHTML = rowHtml
+                }
+            } else {
+                jq("#trace_table").append(rowHtml)
+            }
+        }
+
+        updateOccurredLabels()
+    }
+
+    private fun displayTraceStack(vararg traceStack: TraceSpanInfo) { //todo-chess-equality: [traceStack: List<TraceSpanInfo>]
+        js("portalLog('Displaying trace stack:');")
```

Comment:
Same thing with relying on `js("portalLog(...))`. You have the real portalLog() method already in Kotlin.

In general, you should above `js()` as much as possible.